sds_data_manager/lambda_code/SDSCode/upload_api.py

```diff
@@ -5,40 +5,75 @@
 import os
 
 import boto3
-from imap_data_access import ScienceFilePath
-from imap_data_access import config as imap_data_access_config
-from sqlalchemy import select
-from sqlalchemy.orm import Session
-
-from .database import database as db
-from .database import models
+import botocore
+import imap_data_access
 
 logger = logging.getLogger(__name__)
 logger.setLevel(logging.INFO)
 
-s3 = boto3.client("s3")
+BUCKET_NAME = os.getenv("S3_BUCKET")
+S3_CLIENT = boto3.client("s3")
+
 
+def _file_exists(s3_key_path):
+    """Check if a file exists in the SDS storage bucket at this key."""
+    try:
+        S3_CLIENT.head_object(Bucket=BUCKET_NAME, Key=s3_key_path)
+        # If the head_object operation succeeds, that means there
+        # is a file already at the specified path, so return a 409
+        return True
+    except botocore.exceptions.ClientError:
+        # No file exists
+        return False
 
-def _generate_signed_upload_url(key_path, tags=None):
+
+def _generate_signed_upload_response(s3_key_path, tags=None):
     """Create a presigned url for a file in the SDS storage bucket.
 
-    :param key_path: Required.  A string representing the name of the object to upload.
-    :param tags: Optional.  A dictionary that will be stored in the S3 object metadata.
+    Parameters
+    ----------
+    s3_key_path : str
+        The fully qualified path of the object to upload.
+    tags : dict, optional
+         Additional S3 object metadata to add to the object.
 
-    :return: A URL string if the file was found, otherwise None.
+    Returns
+    -------
+    Response with status code and a pre-signed URL for the object.
     """
-    bucket_name = os.environ["S3_BUCKET"]
-    url = boto3.client("s3").generate_presigned_url(
+    if _file_exists(s3_key_path):
+        # We already have a file at this location, return a 409
+        return {
+            "statusCode": 409,
+            "body": json.dumps(f"{s3_key_path} already exists."),
+            "headers": {
+                "Content-Type": "application/json",
+                "Access-Control-Allow-Origin": "*",
+            },
+        }
+    # We know there isn't an object at this location, so
+    # generate a pre-signed URL for the client to upload to
+    url = S3_CLIENT.generate_presigned_url(
         ClientMethod="put_object",
         Params={
-            "Bucket": bucket_name,
-            "Key": key_path,
+            "Bucket": BUCKET_NAME,
+            "Key": s3_key_path,
             "Metadata": tags or dict(),
         },
         ExpiresIn=3600,
     )
 
-    return url
+    if url is None:
```

Comment:
Everything is checked before this point and returns error messages/codes in the SPICE / Science routines, so it should be good by this point.

In fact, I am not sure that this function can return None... @tech3371 do you know why this was added/needed?
I think it would raise an Exception, and then in the documentation it shows that exception returns a None from a wrapped function, so perhaps that is a bit of an indirection and we don't need it.

https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-presigned-urls.html

I would be happy to remove it, but left things the way they were before for now.

Reply:
Here is the implementation. I'm not sure if `request.url` can be None or not 🤷 
https://github.com/boto/botocore/blob/d85b98d44c3b9ed8661270db60e0b70de85c650d/botocore/signers.py#L323

Reply:
I can't remember why it was added. We can remove this

Reply:
I have removed this in the latest commit.